Change wasm calling convention; set callee from caller.

justinmichaud opened the pull request:
#### c2eee1bb4cec95f71e53400d83a2985463423c17
<pre>
Change wasm calling convention; set callee from caller.
<a href="https://bugs.webkit.org/show_bug.cgi?id=266779">https://bugs.webkit.org/show_bug.cgi?id=266779</a>
<a href="https://rdar.apple.com/120007206">rdar://120007206</a>

Reviewed by Yusuke Suzuki.

This is the first step to running wasm without jit.

This calling convention change opens the door to removing the LLInt entry
thunk by setting the wasm callee from the caller.

When a call is made into wasm, we first create a frame with the JSWebAssemblyFunction
as the callee. From this point on, callees are no longer JSObjects.

The LLInt needs to read stuff from the callee to know what to execute. Today
we have a unique address for each wasm function / tier.

This change means that the LLInt or IPInt can make all functions share an
entrypoint, and there is no more need for the LLInt/IPInt entry thunk.

This patch doesn&apos;t actually remove the thunk yet, it replaces it with a
debug assertion.

JIT tiers don&apos;t need their callee&apos;s to be written this way, and they
should continue to write them themselves. This way, we write the interpreter
callee into our JIT callsites, and never have to worry about attomically
repatching it when we update the entrypoint.

== The path to JITless wasm ==

Once this patch is landed and stable, we can remove the debug assertion
and remove the LLInt entry thunks.

The next step is to start picking off some simple JS-&gt;Wasm thunk cases to replace with
an IPInt-style metadata interpreter + some fixed fast paths. This should save
us some memory initially, and we can eventually remove the entry/exit thunks completely.

== Arm 32 == ( Fixes from Joseph Griego )
n 32-bit we have a totally different CalleeBits representation; it&apos;s morally a
JSValue except it might have the NativeValueTag, so the tag needs to be written
when this slot is initialized, either in the runtime or in JITted code.

AFAICT there&apos;s no existing way to write to a CalleeBits location from C++ right
now so I added some methods in CalleeBits to help with this--it would probably
be better long-term to have a `Register::operator=(CalleeBits)` or something
along those lines.

* JSTests/wasm/stress/cc-int-to-int.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.type.sig_test.func.param.i32.result.i32.table.t.1.funcref.elem.i32.const.0.test.func.test.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.const.42.func.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.call.test.i32.const.1337.func.export.string_appeared_here.param.x.i32.result.i32.local.x.i32.const.98.call_indirect.t.type.sig_test.i32.const.0.i32.add.async test):
* Source/JavaScriptCore/llint/LowLevelInterpreter.cpp:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/offlineasm/arm.rb:
* Source/JavaScriptCore/offlineasm/arm64.rb:
* Source/JavaScriptCore/offlineasm/registers.rb:
* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/runtime/UGPRPair.h:
(JSC::makeUGPRTriple):
(JSC::encodeResult):
(JSC::decodeResult):
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::compileFunction):
* Source/JavaScriptCore/wasm/WasmCalleeGroup.cpp:
(JSC::Wasm::CalleeGroup::CalleeGroup):
* Source/JavaScriptCore/wasm/WasmCalleeGroup.h:
* Source/JavaScriptCore/wasm/WasmCallsiteCollection.cpp:
(JSC::Wasm::CallsiteCollection::updateCallsitesToCallUs):
* Source/JavaScriptCore/wasm/WasmIPIntPlan.cpp:
(JSC::Wasm::IPIntPlan::didCompleteCompilation):
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
(JSC::LLInt::shouldJIT): Deleted.
(JSC::LLInt::jitCompileAndSetHeuristics): Deleted.
(JSC::LLInt::WASM_IPINT_EXTERN_CPP_DECL): Deleted.
(JSC::LLInt::WASM_IPINT_EXTERN_CPP_DECL_1P): Deleted.
(JSC::LLInt::doWasmCall): Deleted.
(JSC::LLInt::doWasmCallIndirect): Deleted.
* Source/JavaScriptCore/wasm/WasmLLIntPlan.cpp:
(JSC::Wasm::LLIntPlan::didCompleteCompilation):
* Source/JavaScriptCore/wasm/WasmOSREntryPlan.cpp:
(JSC::Wasm::OSREntryPlan::work):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::doWasmCall):
(JSC::LLInt::WASM_SLOW_PATH_DECL3):
(JSC::LLInt::doWasmCallIndirect):
* Source/JavaScriptCore/wasm/WasmSlowPaths.h:
* Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
(JSC::Wasm::createJSToWasmWrapper):
* Source/JavaScriptCore/wasm/js/JSToWasm.h:
* Source/JavaScriptCore/wasm/js/WasmToJS.cpp:
(JSC::Wasm::wasmToJS):

Canonical link: <a href="https://commits.webkit.org/273958@main">https://commits.webkit.org/273958@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/41524f158da65ddd6579331a0251eadab23719da

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37234 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16136 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39532 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39799 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33210 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18716 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/13278 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31697 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37797 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13610 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32750 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11857 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11838 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/33397 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41059 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/31110 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33713 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/33831 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37745 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/36942 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12183 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9971 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35888 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13796 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/43811 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8423 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12535 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9007 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/13056 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rkirsling commented on Source/JavaScriptCore/wasm/WasmCalleeGroup.h:
```diff
@@ -191,6 +198,7 @@ class CalleeGroup final : public ThreadSafeRefCounted<CalleeGroup> {
     RefPtr<LLIntCallees> m_llintCallees;
     HashMap<uint32_t, RefPtr<JSEntrypointCallee>, DefaultHash<uint32_t>, WTF::UnsignedWithZeroKeyHashTraits<uint32_t>> m_jsEntrypointCallees;
     FixedVector<CodePtr<WasmEntryPtrTag>> m_wasmIndirectCallEntryPoints;
+    FixedVector<RefPtr<Wasm::Callee>> m_wasmIndirectCallCalees;
```
misspell: `Calees`

rkirsling commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -693,7 +693,7 @@ class B3IRGenerator {
     PartialResult WARN_UNUSED_RETURN addUnreachable();
     PartialResult WARN_UNUSED_RETURN addCrash();
-    PartialResult WARN_UNUSED_RETURN emitIndirectCall(Value* calleeInstance, Value* calleeCode, Value* jsCalleeAnchor, const TypeDefinition&, const Vector<ExpressionType>& args, ResultList&, CallType = CallType::Call);
+    PartialResult WARN_UNUSED_RETURN emitIndirectCall(Value* calleeInstance, Value* calleeCode, Value* calleeCalle, Value* jsCalleeAnchor, const TypeDefinition&, const Vector<ExpressionType>& args, ResultList&, CallType = CallType::Call);
```
misspell: `Calle`

justinmichaud replied:
Haha, Spanish autocorrect

rkirsling reviewed (commented):
This is wonderful! Just pointing out a couple of spelling errors as I'm studying this.

rkirsling commented on Source/JavaScriptCore/interpreter/CalleeBits.h:
```diff
@@ -56,6 +56,13 @@ class CalleeBits {
         return *this;
     }
 
+    static void* boxNativeCalleeOrNull(NativeCallee* callee)
```
I think `boxNativeCalleeIfExists` would be better (otherwise it sounds like you're going to box a null).

rkirsling commented on Source/JavaScriptCore/jit/CCallHelpers.h:
```diff
@@ -962,6 +962,38 @@ class CCallHelpers : public AssemblyHelpers {
+        CCallHelpers::Address calleeFrame = CCallHelpers::Address(MacroAssembler::stackPointerRegister, 0);
+        return calleeFrame.withOffset(CallFrameSlot::callee * static_cast<int>(sizeof(Register)))
+            // calleeFrame is from the caller's perspective
+            .withOffset(-(int) prologueStackPointerDelta());
```
Shouldn't this be a `static_cast` (or `safeCast`)?

rkirsling commented on Source/JavaScriptCore/wasm/WasmFormat.h:
```diff
@@ -749,10 +751,12 @@ struct WasmToWasmImportableFunction {
     using LoadLocation = CodePtr<WasmEntryPtrTag>*;
     static ptrdiff_t offsetOfSignatureIndex() { return OBJECT_OFFSETOF(WasmToWasmImportableFunction, typeIndex); }
     static ptrdiff_t offsetOfEntrypointLoadLocation() { return OBJECT_OFFSETOF(WasmToWasmImportableFunction, entrypointLoadLocation); }
+    static ptrdiff_t offsetOfboxedWasmCalleeLoadLocation() { return OBJECT_OFFSETOF(WasmToWasmImportableFunction, boxedWasmCalleeLoadLocation); }
```
nit: looks like `boxed` needs capitalizing

rkirsling commented on Source/JavaScriptCore/wasm/WasmLLIntPlan.cpp:
```diff
@@ -131,11 +131,15 @@ void LLIntPlan::didCompleteCompilation()
-            jit.store32(CCallHelpers::TrustedImm32(JSValue::NativeCalleeTag), calleeSlot.withOffset(TagOffset));
-#endif
+            {
+                // Debug assertion
```
Is this part still WIP? (Or do we need the non-`good` path to exist?)

justinmichaud replied:
This just adds a debug assertion to check that the callee was set correctly. Eventually we can just remove the whole thunk, once this patch has had an opportunity to get fuzzed

rkirsling commented on Source/JavaScriptCore/wasm/js/WebAssemblyFunctionBase.h:
```diff
@@ -56,6 +57,7 @@ class WebAssemblyFunctionBase : public JSFunction {
     static ptrdiff_t offsetOfSignatureIndex() { return OBJECT_OFFSETOF(WebAssemblyFunctionBase, m_importableFunction) + WasmToWasmImportableFunction::offsetOfSignatureIndex(); }
 
     static ptrdiff_t offsetOfEntrypointLoadLocation() { return OBJECT_OFFSETOF(WebAssemblyFunctionBase, m_importableFunction) + WasmToWasmImportableFunction::offsetOfEntrypointLoadLocation(); }
+    static ptrdiff_t offsetOfboxedWasmCalleeLoadLocation() { return OBJECT_OFFSETOF(WebAssemblyFunctionBase, m_importableFunction) + WasmToWasmImportableFunction::offsetOfboxedWasmCalleeLoadLocation(); }
```
nit: another identifier which needs `boxed` capitalized

justinmichaud commented on Source/JavaScriptCore/wasm/js/WebAssemblyFunction.h:
```diff
@@ -92,6 +92,11 @@ class WebAssemblyFunction final : public WebAssemblyFunctionBase {
     CodePtr<WasmEntryPtrTag> m_jsEntrypoint;
+    // This is the callee needed by LLInt/IPInt
+    // This keeps it alive
+    RefPtr<Wasm::Callee> m_wasmCallee;
```
1) Does this need to keep the LLInt callee alive?
2) If so, is this going to be a problem? Should we clear this out when we tier up?

justinmichaud replied:
I'll remove this

Constellation reviewed (approved):
r=me

Commit:
Change wasm calling convention; set callee from caller.
https://bugs.webkit.org/show_bug.cgi?id=266779
rdar://120007206

Reviewed by Yusuke Suzuki.

This is the first step to running wasm without jit.

This calling convention change opens the door to removing the LLInt entry
thunk by setting the wasm callee from the caller.

When a call is made into wasm, we first create a frame with the JSWebAssemblyFunction
as the callee. From this point on, callees are no longer JSObjects.

The LLInt needs to read stuff from the callee to know what to execute. Today
we have a unique address for each wasm function / tier.

This change means that the LLInt or IPInt can make all functions share an
entrypoint, and there is no more need for the LLInt/IPInt entry thunk.

This patch doesn't actually remove the thunk yet, it replaces it with a
debug assertion.

JIT tiers don't need their callee's to be written this way, and they
should continue to write them themselves. This way, we write the interpreter
callee into our JIT callsites, and never have to worry about attomically
repatching it when we update the entrypoint.

== The path to JITless wasm ==

Once this patch is landed and stable, we can remove the debug assertion
and remove the LLInt entry thunks.

The next step is to start picking off some simple JS->Wasm thunk cases to replace with
an IPInt-style metadata interpreter + some fixed fast paths. This should save
us some memory initially, and we can eventually remove the entry/exit thunks completely.

== Arm 32 == ( Fixes from Joseph Griego )
n 32-bit we have a totally different CalleeBits representation; it's morally a
JSValue except it might have the NativeValueTag, so the tag needs to be written
when this slot is initialized, either in the runtime or in JITted code.

AFAICT there's no existing way to write to a CalleeBits location from C++ right
now so I added some methods in CalleeBits to help with this--it would probably
be better long-term to have a `Register::operator=(CalleeBits)` or something
along those lines.

* JSTests/wasm/stress/cc-int-to-int.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.type.sig_test.func.param.i32.result.i32.table.t.1.funcref.elem.i32.const.0.test.func.test.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.i32.const.42.func.export.string_appeared_here.param.x.i32.result.i32.i32.add.local.x.call.test.i32.const.1337.func.export.string_appeared_here.param.x.i32.result.i32.local.x.i32.const.98.call_indirect.t.type.sig_test.i32.const.0.i32.add.async test):
* Source/JavaScriptCore/llint/LowLevelInterpreter.cpp:
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/offlineasm/arm.rb:
* Source/JavaScriptCore/offlineasm/arm64.rb:
* Source/JavaScriptCore/offlineasm/registers.rb:
* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/runtime/UGPRPair.h:
(JSC::makeUGPRTriple):
(JSC::encodeResult):
(JSC::decodeResult):
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::compileFunction):
* Source/JavaScriptCore/wasm/WasmCalleeGroup.cpp:
(JSC::Wasm::CalleeGroup::CalleeGroup):
* Source/JavaScriptCore/wasm/WasmCalleeGroup.h:
* Source/JavaScriptCore/wasm/WasmCallsiteCollection.cpp:
(JSC::Wasm::CallsiteCollection::updateCallsitesToCallUs):
* Source/JavaScriptCore/wasm/WasmIPIntPlan.cpp:
(JSC::Wasm::IPIntPlan::didCompleteCompilation):
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
(JSC::LLInt::shouldJIT): Deleted.
(JSC::LLInt::jitCompileAndSetHeuristics): Deleted.
(JSC::LLInt::WASM_IPINT_EXTERN_CPP_DECL): Deleted.
(JSC::LLInt::WASM_IPINT_EXTERN_CPP_DECL_1P): Deleted.
(JSC::LLInt::doWasmCall): Deleted.
(JSC::LLInt::doWasmCallIndirect): Deleted.
* Source/JavaScriptCore/wasm/WasmLLIntPlan.cpp:
(JSC::Wasm::LLIntPlan::didCompleteCompilation):
* Source/JavaScriptCore/wasm/WasmOSREntryPlan.cpp:
(JSC::Wasm::OSREntryPlan::work):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::doWasmCall):
(JSC::LLInt::WASM_SLOW_PATH_DECL3):
(JSC::LLInt::doWasmCallIndirect):
* Source/JavaScriptCore/wasm/WasmSlowPaths.h:
* Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
(JSC::Wasm::createJSToWasmWrapper):
* Source/JavaScriptCore/wasm/js/JSToWasm.h:
* Source/JavaScriptCore/wasm/js/WasmToJS.cpp:
(JSC::Wasm::wasmToJS):

Canonical link: https://commits.webkit.org/273958@main
